### PR TITLE
Fix/improve handling of forward declarations in nimsuggest

### PR DESCRIPTION
**Pull request by yyoncho**
- ideUse now works fine when invoked on the implementation
- implemented ideDeclaration to make cover lsp feature textDocument/declaration
- fixed performance issue related to deduplicating symbols. Now the
deduplication happens after the symbols are filtered. As a alternative we might
change the way cached symbols are stored(e. g. use set).
- I also fixed the way globalSymbols work. Now it will sort the responses based
on the match location to make sure that the results are sorted in user friendly way.

### CODE REVIEW

File: nimsuggest/nimsuggest.nim [hunk 1]
```diff
@@ -852,16 +878,57 @@ proc executeNoHooksV3(cmd: IdeCmd, file: AbsoluteFile, dirtyfile: AbsoluteFile,
...
+      if leftIndex == rightIndex:
+        result = cmp(toLowerAscii(leftString),
+                     toLowerAscii(rightString))
+      else:
+        result = cmp(leftIndex, rightIndex)
+
+    # ... and send first 100 results
+    if res.len > 0:
+      for s in res[0 .. min(100, res.len - 1)]:
+        graph.suggestResult(s.sym, s.info)
```
Comment (Araq) [inline nimsuggest/nimsuggest.nim]:
```suggestion
      for i in 0 .. min(100, res.len - 1):
        let s = res[i]
        graph.suggestResult(s.sym, s.info)
```

**Comment by Araq** [comment]
Not sure if it's related but on OSX tests/threads/t7172.nim fails.

**Comment by yyoncho** [comment]
> Not sure if it's related but on OSX tests/threads/t7172.nim fails.

it is good after the merge, I could rebase if you want.

**Comment by yyoncho** [comment]
@narimiran can you backport this one?

**Comment by narimiran** [comment]
> @narimiran can you backport this one?

Done!

**Pull request merged**